src/adapter/rust_type_name.rs

```diff
@@ -0,0 +1,464 @@
+                    .map(|(name, ty)| Arg::Named(name, Type(ty)))
+                    .chain(fnp.sig.is_c_variadic.then_some(Arg::Dots)),
+                |arg, f| match arg {
+                    Arg::Named(name, ty) => write!(f, "{name}: {ty}"),
+                    Arg::Dots => write!(f, "..."),
+                },
+            )?;
+
+            write!(f, ")")?;
+
+            if let Some(output) = &fnp.sig.output {
+                write!(f, " -> {}", Type(output))?;
+            }
+
+            Ok(())
+        }
+        rustdoc_types::Type::Tuple(tys) => {
+            write!(f, "(")?;
+            intersperse_with(f, ", ", tys, Type)?;
+            write!(f, ")")
+        }
+        rustdoc_types::Type::Slice(ty) => write!(f, "[{}]", Type(ty)),
+        // according to docs, `len` is not guaranteed to roundtrip
+        rustdoc_types::Type::Array { type_, len } => write!(f, "[{}; {}]", Type(type_), len),
+        rustdoc_types::Type::ImplTrait(vec) => {
+            write!(f, "impl ")?;
+            intersperse_with(f, " + ", vec, GenericBound)
+        }
+        rustdoc_types::Type::Infer => write!(f, "_"),
+        rustdoc_types::Type::RawPointer { is_mutable, type_ } => {
+            let kind = if *is_mutable { "mut" } else { "const" };
+
+            write!(f, "*{kind} {}", Type(type_))
+        }
+        rustdoc_types::Type::BorrowedRef {
+            lifetime,
+            is_mutable,
+            type_,
+        } => {
+            write!(f, "&")?;
+            if let Some(lt) = lifetime {
+                write!(f, "{lt} ")?;
+            }
+
+            if *is_mutable {
+                write!(f, "mut ")?;
+            }
+
+            write!(f, "{}", Type(type_))
+        }
+        rustdoc_types::Type::QualifiedPath {
+            name,
+            args,
+            self_type,
+            trait_,
+        } => {
+            if let Some(trait_) = trait_ {
+                write!(f, "<{} as {}>", Type(self_type), Path(trait_))?;
+            } else {
+                write!(f, "{}", Type(self_type))?;
+            }
+
+            write!(f, "::{}{}", name, GenericArgs(args))
+        }
+        rustdoc_types::Type::Pat { .. } => unimplemented!("Type::Pat is unstable"),
+    }
+}
+
+display_wrapper!(Type, fmt_type);
+
+fn fmt_generic_bound(this: &GenericBound, f: &mut Formatter<'_>) -> Result {
+    match &this.0 {
+        rustdoc_types::GenericBound::TraitBound {
+            trait_,
+            generic_params,
+            modifier,
+        } => {
+            match modifier {
+                rustdoc_types::TraitBoundModifier::Maybe => write!(f, "?")?,
+                // TODO: check this/find a good reference.  it's currently unstablke
+                rustdoc_types::TraitBoundModifier::MaybeConst => write!(f, "~const ")?,
+                rustdoc_types::TraitBoundModifier::None => (),
+            };
+
+            write!(f, "{}", Path(trait_))?;
+
+            if generic_params.is_empty() {
```

Comment:
Should this be a `!is_empty()` instead?

Consider writing a test case to trigger this bug before fixing it, since it's a sneaky one.

Reply:
Fixed, this will be caught by e.g. `impl for<'a> Fn(&'a ())` which is tested.